### PR TITLE
[#IOPAE-140] New Manage Flow Middleware

### PR DESCRIPTION
**Pull request by giuseppedipinto**
New middleware for MANAGE Flow

#### List of Changes

Added new AzureUserAttributesManageMiddleware

#### Motivation and Context

In order to use *Manage Api Key* for *Services Create/Update*, we need a new middleware for checking user permissions

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by balanza — CHANGES_REQUESTED** [review]
I see some macro-issue in these changes:

1. **There are no tests** - thus is hard to tell if the solution works, if the exposed API is convenient, and how it performs on edge cases
1. **code is 90% duplicated** because we already have a middleware that checks on subscriptions. 
1. **ServiceModel is an unused dependency** for the manage branch, nevertheless is required by the middleware
1. (maybe not an issue) **this middleware works at a different level of abstraction**: it has the concept of USE/MANAGE flow, while other middlewares only reason on raw data

The first issue I think is self explanatory. About the second, try to figure the following possible future scenarios:
* _we need to fix a bug in how we check subscriptions_ - now we have two disjoint places to fix, test and deploy
* _we must allow manage key only_ - how would you?

The third comes with the second, and the latter is maybe just me ranting.

----
**Proposal**
What if we have:
* A middleware that checks on manage subscriptions only
* A _meta-_ middleware that takes two (or more) middlewares, evaluate them in sequence until one is satisfied (example name: `sequenceMiddleware`). I'm thinking on an API like:
```ts
sequenceMiddleware(
  AzureUserAttributesMiddleware(ServiceModel),
  AzureUserAttributesManageMiddleware
)
```

The benefit would be:
* to have a well-scoped middleware for the specific case (`AzureUserAttributesManageMiddleware`)
* to let the implementing functions decide which logic to apply
* to add a reusable feature for others (`sequenceMiddleware`)

File: src/utils/middlewares/azure_user_attributes_manage.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+>;
+
+/**
+ * A middleware that will extract custom user attributes from the request, that support also **MANAGE** flow.
+ *
+ * The middleware expects the following headers:
+ *
+ *   x-subscription-id:     The user's subscription id.
+ *
+ * In **LEGACY flow** is used to retrieve the associated Service.
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply src/utils/middlewares/azure_user_attributes_manage.ts:50]: Actually, you understood correctly at the beginning, that with **LEGACY flow** we mean the current flow, and with **MANAGE flow** we mean the new flow. 
So in this sense, it is the **LEGACY** that would have been completely replaced by the **MANAGE**.

File: src/utils/middlewares/azure_user_attributes_manage.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    };
+
+    return E.right<
+      IResponse<
+        | "IResponseErrorForbiddenNotAuthorized"
+        | "IResponseErrorQuery"
+        | "IResponseErrorInternal"
+      >,
+      IAzureUserAttributes
+    >(authInfo);
```
[the remark this answers is not in the document]
Comment (giuseppedipinto) [reply src/utils/middlewares/azure_user_attributes_manage.ts:149]: Unfortunately we cannot fully reference the `azure_user_attributes` module because within this module there are controls that would not work with the _**Manage**_ flow and the main goal we set ourselves was not to touch `azure_user_attributes` in order to avoid potential impacts on many APIs using this middleware.
For this reason we agreed to **temporarily duplicate** `azure_user_attributes` by adding an _"else"_ condition that would **temporarily** make both flows possible.
So summarizing, the answer is "yes, the new middleware duplicates `azure_user_attributes` by adding a condition (else) in which it also manages the MANAGE flow.

**Review by balanza — APPROVED** [review]
